src/IRM/IRMAdaptiveCurve.sol

```diff
@@ -0,0 +1,208 @@
+/// Returned rates are extended to RAY per second to be compatible with the EVK.
+contract IRMAdaptiveCurve is IIRM {
+    /// @dev Unit for internal precision.
+    int256 internal constant WAD = 1e18;
+    /// @dev Unit for internal precision.
+    int256 internal constant YEAR = int256(365.2425 days);
+    /// @notice The name of the IRM.
+    string public constant name = "IRMAdaptiveCurve";
+    /// @notice The utilization rate targeted by the model.
+    /// @dev In WAD units.
+    int256 public immutable TARGET_UTILIZATION;
+    /// @notice The initial interest rate at target utilization.
+    /// @dev In WAD per second units.
+    /// When the IRM is initialized for a vault this is the rate at target utilization that is assigned.
+    int256 public immutable INITIAL_RATE_AT_TARGET;
+    /// @notice The minimum interest rate at target utilization that the model can adjust to.
+    /// @dev In WAD per second units.
+    int256 public immutable MIN_RATE_AT_TARGET;
+    /// @notice The maximum interest rate at target utilization that the model can adjust to.
+    /// @dev In WAD per second units.
+    int256 public immutable MAX_RATE_AT_TARGET;
+    /// @notice The steepness of the interest rate line.
+    /// @dev In WAD units.
+    int256 public immutable CURVE_STEEPNESS;
+    /// @notice The speed at which the rate at target is adjusted up or down.
+    /// @dev In WAD per second units.
+    /// For example, with `2e18 / 24 hours` the model will 2x `rateAtTarget` if the vault is fully utilized for a day.
+    int256 public immutable ADJUSTMENT_SPEED;
+
+    /// @notice Internal cached state of the interest rate model.
+    struct IRState {
+        /// @dev The current rate at target utilization.
+        uint208 rateAtTarget;
+        /// @dev The timestamp of the last update to the model.
+        uint48 lastUpdate;
+    }
+
+    /// @notice Get the internal cached state of a vault's irm.
+    mapping(address => IRState) internal irState;
+
+    error InvalidParams();
+
+    /// @notice Deploy IRMAdaptiveCurve.
+    /// @param _TARGET_UTILIZATION The utilization rate targeted by the interest rate model.
+    /// @param _INITIAL_RATE_AT_TARGET The initial interest rate at target utilization.
+    /// @param _MIN_RATE_AT_TARGET The minimum interest rate at target utilization that the model can adjust to.
+    /// @param _MAX_RATE_AT_TARGET The maximum interest rate at target utilization that the model can adjust to.
+    /// @param _CURVE_STEEPNESS The steepness of the interest rate line.
+    /// @param _ADJUSTMENT_SPEED The speed at which the rate at target utilization is adjusted up or down.
+    constructor(
+        int256 _TARGET_UTILIZATION,
+        int256 _INITIAL_RATE_AT_TARGET,
+        int256 _MIN_RATE_AT_TARGET,
+        int256 _MAX_RATE_AT_TARGET,
+        int256 _CURVE_STEEPNESS,
+        int256 _ADJUSTMENT_SPEED
+    ) {
+        // Validate parameters.
+        if (_TARGET_UTILIZATION < 0 || _TARGET_UTILIZATION > 1e18) {
```

Comment:
IIUC, a 50% utilisation implies no kink, and utilisation < 50% results in the segment to the left of the kink being steeper than the right.

No action necessary, but it seems like this formulation may not work as expected with low utilisations.

Reply:
I'm wondering if we should allow a configuration with inverted slopes. Imo this breaks a key property of linear kink IRMs.

Reply:
I did the math,

$$\text{targetUtilization}>\frac{1}{\text{curveSteepness}+1}$$

is the actual cut-off point where slopes invert:
- At 10%, steepness < 9.9
- At 20%, steepness < 2
- At 30%, steepness < 2.33
- At 40%, steepness < 1.5
- At 50%, steepness < 1

If we wanted to block this behavior we could check the inequality.

Reply:
Well, there are probably several ways to create a bad IRM instance. As Darek says, "many dumb ways to die". Since we are eliminating most of these by choosing good parameters off-chain, it's probably overkill to detect it in the constructor. It's good to be aware of it though, since I could certainly see someone requesting an adaptive IRM that targets 50%.